### PR TITLE
feat: Enable HMR for React components

### PR DESCRIPTION
**Pull request by ElMassimo**
Hi! The site looks very nice!

Felt a good time to do a contribution now that it's out in the open 😃 

### Description 📖 

This pull request enables HMR for React components, in order to get the best out of Vite.

### Background 📜 

<kbd>[@vitejs/plugin-react-refresh](https://www.npmjs.com/package/@vitejs/plugin-react-refresh)</kbd> is a core Vite plugin that enables React Refresh in Vite projects.

There is a [shortcoming](https://github.com/vitejs/vite/issues/1984#issuecomment-778289660) in this plugin when using it in [non-HTML entrypoints](https://vite-ruby.netlify.app/guide/troubleshooting.html#hot-module-refresh-does-not-work-for-react), which is why Vite Ruby provides a <kbd>[vite_react_refresh_tag](https://vite-ruby.netlify.app/guide/rails.html#enabling-hot-module-reload-%F0%9F%94%A5)</kbd> helper to make things easier for the end user.

### Screenshot 📷 

Changes are now reflected without reloading the page.

https://user-images.githubusercontent.com/1158253/116835215-ed06f680-ab97-11eb-8825-56d7126e4302.mp4

### CODE REVIEW

File: app/serializers/season_serializer.rb [hunk 1]
```diff
@@ -3,8 +3,8 @@
 # Serializes the season to JSON
 class SeasonSerializer < Oj::Serializer
   attributes :id, :name, :season_number, :episode_count, :slug
-  serializer_attributes :poster_url
 
+  attribute \
   def poster_url
     season.poster.url
   end
```
Comment (ElMassimo) [inline app/serializers/season_serializer.rb:10]: This is unrelated, just wanted to mention that you can do this in Oj Serializers 😄

**Comment by maxjacobson** [comment]
Whoa niceee.

I haven't wired up CI at all bc I haven't written any tests at all and have just been working on my own. The one thing I do have so far is `exe/lint`, which I'd prefer to keep green before merging this in. Otherwise it looks great and I appreciate you helping out 🤩

**Comment by maxjacobson** [comment]
Appreciate it!!

While I have you, here are a few quick questions about vite and oj serializers! 

1. I've seen that things usually seem to work fine whether I run `bin/vite dev` or not, but that they seem to work _better_ if I do. Is that right?

    I'm asking in part because, every now and then, my rails server kind of hangs, and won't even respond to `ctrl c`. I've recently stopped running `bin/vite dev` because that makes it stop occasionally hanging, but it also makes the experience a little worse, e.g. the live code reloading stops working.

    When it hangs, it seems to hang for about a minute, and I think it may have been waiting on this timeout (i spent a few minutes trying to debug it before giving up):
    https://github.com/ncr/rack-proxy/blob/bee97d2b25a8c6611b4dccba2bb62288f5ac3596/lib/rack/proxy.rb#L52
    
    And I see things like this in the rails logs:
    
    ```
    2021-05-03 18:16:10 -0400 Rack app ("GET /vite-dev/@fs/Users/max.jacobson/src/gh/maxjacobson/seasoning/app/javascript/components/YourShowsList.tsx" - (127.0.0.1)): #<Net::OpenTimeout: execution expired>
    ```
    
    I found myself craving a simpler integration which _requires_ you to run `bin/vite dev` and loads assets directly from the `localhost:3036` URL, rather than proxying thru the rails server. I'm curious if you considered that, or if it's possible to configure it to do that. I see that it is possible to set `autoBuild` to false, which seems to disable the auto-building behavior, but still loads the assets via the rack proxying, which sometimes times out for me, for reasons unknown 🤔 
2. I've seen that, occasionally, after making changes to an OJ serializer, i need to stop and start the server for the changes to take effect. Is that a known issue?
3. is there any way with OJ serializers to pass in multiple objects to an instance of a serializer? I was kind of craving that, and I worked around it by introducing a new class, which wraps the two things, which is fine, but curious if there's another way to do it:
    https://github.com/maxjacobson/seasoning/blob/8ac9c91a031e05293839ce8297f23a952397b800/app/controllers/api/profiles_controller.rb#L12
    https://github.com/maxjacobson/seasoning/blob/8ac9c91a031e05293839ce8297f23a952397b800/app/models/profile.rb#L3



No presh to answer these, but since you swung by, these popped into my head. No good deed goes unpunished 😈

**Comment by ElMassimo** [comment]
Haha, sure, happy to answer a few.

1. I haven't experienced timeouts—it seems unusual that a request to the Vite dev server hangs forever. If the 
issue has something to do with Vite, circumventing the proxy might not solve the problem.

I'd say, try running Vite with logging on and next time the problem occurs check what's the last thing 
that gets printed (who knows might be related with React, or perhaps it would go away now that the plugin 
enables HMR).

To show logs, you could run `bin/vite dev --debug`, but since this seems a lower-level error, better get non-Vite logs as well by running `DEBUG=* bin/vite dev`.

2. Code inside the `app/serializers` folder should be auto-reloaded.

This library was extracted after a few years of usage at Cipher—can't recall a similar issue happening to anyone. Caching can cause a similar result, but I noticed the app is not using it, so we can discard that.

What kind of changes are not being reflected? Adding attributes for example? If you add a `binding.pry` in a serializer attribute does it pause execution?

3. Adding a "glue object" like `Profile` there is a valid solution.

An alternative that we used (sparingly), is to pass [options](https://github.com/ElMassimo/oj_serializers/blob/master/lib/oj_serializers/serializer.rb#L35-L39) to the serializer.

Using your example:

```ruby
render json: {
  profile: ProfileSerializer.one(human, viewer: current_human),
}
```

and in the serializer:

```ruby
def viewer
  options.fetch(:viewer)
end
```

**Comment by ElMassimo** [comment]
Regarding question 1, you could also [run `bundle open vite_rails`](https://maximomussini.com/posts/debugging-ruby-libraries/) and modify [this line](https://github.com/ElMassimo/vite_ruby/blob/main/vite_rails/lib/vite_rails/engine.rb#L7) to pass something such as `read_timeout: 1`.

Lowering to one second might make this issue less annoying when it does happen, or easier to debug.

Ideally, we get to the root cause—otherwise I might provide a way to configure it in Vite Ruby, such as exposing the timeout in the config options, or by using a default much lower than 60 seconds.

**Comment by ElMassimo** [comment]
[discussion]: https://discord.com/channels/804011606160703521/804081869523124274/848178143012323351

@maxjacobson In case it's useful, a similar [discussion] popped up recently in the Discord server, if ⓵ is still happening you could try adding the following to your `~/.zshrc` file:

```
ulimit -n 10240
```

That [workaround][discussion] would imply that the error of Vite not responding _would_ be related to [file descriptor limits](https://wilsonmar.github.io/maximum-limits/) in the OS, which makes sense given how fast HMR can happen and the fact that it could trigger a lot of file reads.

Let me know if that helps!

**Comment by maxjacobson** [comment]
@ElMassimo hey! thanks for all that info. I will try that workaround if I hit the issue again. I've been slow to reply mainly because I haven't had time to really work on this recently. Just wanted to acknowledge that I appreciate you :))

BTW: Would be happy to lurk in the discord but not sure how to join! I am a bit of a discord noob

**Comment by maxjacobson** [comment]
@ElMassimo amazing!!
